# Ichthyosis RNA-Seq Analysis (STAR + Charactaristic Direction)
RNA-Seq analysis of Ichthyosis vs. Control samples.
Prior to the analysis, the data was mapped by using STAR.
This code includes steps of:
  * Quantification - by using featureCounts
  * Filteration & normaliztion steps
  * Differntial Expression Analysis - by using Characteristic Direction

```{r include=FALSE}
source ("https://bioconductor.org/biocLite.R")
biocLite()
require(Rsubread)
require(knitr)
```

## Pre-processing of the data
First, we would like to set the directory and get access to the BAM files which were the output of the mapping step with STAR

```{r}
# set the directory
knitr::opts_knit$set(root.dir='/Volumes/ShaiM/Netherton/Shai')
# get all the BAM files
bam_files <- list.files(path=".", pattern = ".bam")
bam_files
```

## Quantification by using featureCounts
Now we would like to quantify the data in order to check how many reads are assigned to each gene

```{r}
for (bam_file in bam_files) {
  bam_counted_features <- Rsubread::featureCounts(files = bam_file, 
                                                  annot.ext = "/Users/shai/Desktop/STAR/Homo_sapiens.GRCh38.87.gtf",
                                                  isGTFAnnotationFile = T,
                                                  GTF.featureType = "exon",
                                                  GTF.attrType = "gene_name")
  # Extract the counts object from the list
  extracted_counts <- bam_counted_features$counts
  extracted_counts_genes <- rownames(bam_counted_features$counts)
  extracted_counts_numbers <- bam_counted_features$counts
  dim(bam_counted_features$counts)
  raw_data <- data.frame(extracted_counts_genes, extracted_counts_numbers)
  # Extract sample name
  sample_name <- strsplit(bam_file, '.', fixed=TRUE)[[1]][1]
  # Add .txt extension
  sample_name_txt <- paste0(sample_name, 'extractedCounts.txt')
  # Write the extracted_counts object to the .txt file
  write.table(raw_data, file=sample_name_txt, sep="\t")
}

txt_files <- c(list.files(path=".", pattern="Counts.txt"))
txt_files
```


## Filtration and normalization steps
We want to prepare the raw data for the diffrential expression analysis

```{r}
concatenated_dataframe <- as.data.frame(matrix(nrow=0, ncol=3))

# Loop through TXT files
for (txt_file in txt_files)
{
  # Extract sample name
  sample_name <- gsub('extractedCounts.txt', '', txt_file)
  
  # Get temporary dataset
  temp_dataset <- read.table(txt_file, header=T, sep="\t")
  
  # Rename column
  colnames(temp_dataset)[2] <- 'readcount'
  
  # Add 'sample_name' column
  temp_dataset$sample_name <- sample_name
  
  # Add to concatenated dataframe
  concatenated_dataframe <- rbind(concatenated_dataframe, temp_dataset)
}
head(concatenated_dataframe)

# Cast matrix
count_matrix <- reshape2::dcast(extracted_counts_genes ~ sample_name, value.var = 'readcount', data=concatenated_dataframe)


head(count_matrix)
rownames(count_matrix) <- count_matrix[,1]
count_matrix[,1]<-NULL
dim(count_matrix)
count_matrix <- count_matrix[apply(count_matrix==0,1,sum)<=10,]
dim(count_matrix)
colnames(count_matrix) <- gsub('_L001_R1_001','',colnames(count_matrix))
colnames(count_matrix) <- gsub('_L002_R1_001','',colnames(count_matrix))

write.table(count_matrix, file='big_matrix.txt', sep="\t", quote = F)
```

In the next step, we will leave R and move to **Python**

This is the code that we will run in Python:
  
  `import pandas as pd`
  `def main():`
  `from clustergrammer import Network`
	`net = Network()`
	`net.load_file('big_matrix.txt')`
	`# net.make_clust(dist_type='cos',views=['N_row_sum', 'N_row_var'])`

	`net.filter_N_top('row',500, rank_type='sum')`

	`net.normalize(axis='row', norm_type='zscore')`
	`net.write_matrix_to_tsv('small_matrix.txt')`
  `main()`

The `small_matrix.txt` now needs to be uploaded to the **clustergrammer** to check for any clusters
we will upload the `small_matrix.txt` back to R 

```{r}
count_matrix <- read.table("small_matrix.txt")
```



```{r}

colnames(count_matrix) <- as.list(c("1","1","0","0","0","0","0","0","0","0","0","0","1","1","1","1","0","1"))

header <- colnames(count_matrix)
genes <- as.vector(rownames(count_matrix))
mat <- as.matrix(count_matrix[1:nrow(count_matrix),1:ncol(count_matrix)])

length(header)
length(genes)
dim(mat)

ctrlMat <- mat[,header==0]
expmMat<- mat[,header==1]
colnames(ctrlMat) <- as.list(c("0","0","0","0","0","0","0","0","0","0","0"))
colnames(expmMat) <- as.list(c("1","1","1","1","1","1","1"))
head(ctrlMat)
head(expmMat)
dim(ctrlMat)
dim(expmMat)
count_matrix <- cbind(ctrlMat,expmMat)
dim(count_matrix)
head(count_matrix)
```


## Differential Expression Analysis (Characteristic Direction)
Here are the functions we will need to use 

```{r}
"nipals" <-function(X,a,it=10,tol=1e-4) 
    #fct nipals calculates the principal components of a given data matrix X according to 
    #the NIPALS algorithm (Wold).
    #X...data matrix, a...number of components, 
    #it...maximal number of iterations per component,
    #tol...precision tolerance for calculation of components
  {
    Xh <- scale(X,center=TRUE,scale=FALSE)		#mean-centering of data matrix X
    nr <- 0
    T <- NULL
    P <- NULL
    pcvar <- NULL
    varTotal <- sum(diag(var(Xh)))
    currVar <- varTotal
    
    for (h in 1:a){
      th <- Xh[,1]		#starting value for th is 1st column of Xh
      ende <- FALSE
      #3 inner steps of NIPALS algorithm
      while (!ende){
        nr <- nr+1
        
        # the result of matrix multiplication operation (%*%) is a matrix of a single
        # valule. A matrix cannot multiply another using scalar multiplication (*).
        # as.vector convert a value of class matrix to a value of class double.
        # (A'*B)' = B'*A
        ph <- t((t(th)%*%Xh) * as.vector(1/(t(th)%*%th)))	#LS regression for ph
        ph <- ph * as.vector(1/sqrt(t(ph)%*%ph))		#normalization of ph
        thnew <- t(t(ph)%*%t(Xh) * as.vector(1/(t(ph)%*%ph)))	#LS regression for th
        prec <- t(th-thnew)%*%(th-thnew)	#calculate precision
        # cat("actual precision: ",sqrt(prec),"\n")
        th <- thnew	#refresh th in any case
        #check convergence of th
        if (prec <= (tol^2)) {
          ende <- TRUE
        }
        else if (it <= nr) {	#too many iterations
          ende <- TRUE
          cat("\nWARNING! Iteration stop in h=",h," without convergence!\n\n")
        }
      }
      Xh <- Xh-(th%*%t(ph))	#calculate new Xh
      T <- cbind(T,th)	#build matrix T
      P <- cbind(P,ph)	#build matrix P
      oldVar <- currVar
      currVar <- sum(diag(var(Xh)))
      pcvar <- c(pcvar,(oldVar-currVar)/varTotal)
      nr <- 0
    }
    list(T=T,P=P,pcvar=pcvar)
  }


"chdir" <-function(ctrl,expm,genes,r=1)
    # This function caclulates the characteristic direction for a gene expression dataset.
    #  	ctrl: control gene expressoion data, a matrix object
    #  	expm: experiment gene expression data, a matrix object
    #  	b: return value, a vector of n-components, representing the characteristic
    #          direction of the gene expression dataset. n equals to the number of genes in the 
    #          expression dataset. b is also a matrix object. b is sorted by its components' 
    #          absolute values in descending order.
    #  	r: regularized term. A parameter that smooths the covariance matrix and reduces
    #          potential noise in the dataset. The default value for r is 1, no regularization.
    #
    #       For the input matrix rows are genes and columns are gene expression profiles.
#       r is the regulization term ranging [0,1]. b is the characteristic direction.
#       ctrl(control) and expm(experiment) matrices should have the same number
#       of genes(rows). 
#
#       Author: Qiaonan Duan
#       Ma'ayan Lab, Icahn School of Medicine at Mount Sinai
#       Jan.13, 2014
#
#		Add gene symbols to results. Apr. 4, 2014

  {
    
    if(dim(ctrl)[1]!=dim(expm)[1]){
      stop('Control expression data must have equal number of genes as experiment expression data!')
    }
    
    if(any(is.na(ctrl))||any(is.na(expm))){
      stop('Control expression data and experiment expression data have to be real numbers. NA was found!')
    }
    
    
    # There should be variance in expression values of each gene. If  
    # gene expression values of a gene are constant, it would dramatically
    # affect the LDA caculation and results in a wrong answer.
    constantThreshold <- 1e-5;
    ctrlConstantGenes <- diag(var(t(ctrl))) < constantThreshold
    expmConstantGenes <- diag(var(t(expm))) < constantThreshold
    
    if (any(ctrlConstantGenes)){
      errMes <- sprintf('%s row(s) in control expression data are constant. Consider Removing the row(s).',paste(as.character(which(ctrlConstantGenes)),collapse=','))
      stop(errMes)
    }else if(any(expmConstantGenes)){
      errMes <- sprintf('%s row(s) in experiment expression data are constant. Consider Removing the row(s).',paste(as.character(which(expmConstantGenes)),collapse=','))
      stop(errMes)
    }
    
    # place control gene expression data and experiment gene expression data into
    # one matrix
    combinedData <- cbind(ctrl,expm)
    
    # get the number of samples, namely, the total number of replicates in  control 
    # and experiment. 
    dims <- dim(combinedData)
    samplesCount <- dims[2]
    
    # the number of output components desired from PCA. We only want to calculate
    # the chdir in a subspace that capture most variance in order to save computation 
    # workload. The number is set 20 because considering the number of genes usually 
    # present in an expression matrix 20 components would capture most of the variance.
    componentsCount <- min(c(samplesCount-1,20))
    
    
    # use the nipals PCA algorithm to calculate R, V, and pcvars. nipals algorithm
    # has better performance than the algorithm used by R's builtin PCA function.
    # R are scores and V are coefficients or loadings. pcvars are the variances 
    # captured by each component 
    pcaRes <- nipals(t(combinedData),componentsCount,1e5,1e-4)
    R <- pcaRes$T
    V <- pcaRes$P
    pcvars <- pcaRes$pcvar
    
    
    # we only want components that cpature 95% of the total variance or a little above.
    # cutIdx is the index of the compoenent, within which the variance is just equal
    # to or a little greater than 95% of the total.
    cutIdx <- which(cumsum(pcvars)>0.95)
    if(length(cutIdx)==0){
      cutIdx <- componentsCount
    }else{
      cutIdx <- cutIdx[1]
    }
    
    # slice R and V to only that number of components.
    R <- R[,1:cutIdx]
    V <- V[,1:cutIdx]
    
    # the difference between experiment mean and control mean.
    meanvec <- rowMeans(expm) - rowMeans(ctrl)
    
    
    # all the following steps calculate shrunkMats. Refer to the ChrDir paper for detail.
    # ShrunkenMats are the covariance matrix that is placed as denominator 
    # in LDA formula. Notice the shrunkMats here is in the subspace of those components
    # that capture about 95% of total variance.
    Dd <- t(R)%*%R/samplesCount
    Dd <- diag(diag(Dd))
    sigma <- mean(diag(Dd))
    shrunkMats <- r*Dd + sigma*(1-r)*diag(dim(R)[2])
    
    # The LDA formula.
    #  V%*%solve(shrunkMats)%*%t(V) transforms the covariance matrix from the subspace to full space.
    b <- V%*%solve(shrunkMats)%*%t(V)%*%meanvec
    
    # normlize b to unit vector
    b <- b*as.vector(sqrt(1/t(b)%*%b))
    
    # sort b to by its components' absolute value in decreasing order and get the 
    # sort index
    sortRes <- sort(abs(b),decreasing=TRUE,index.return=TRUE)
    
    # sort b by the sort index
    bSorted <- as.matrix(b[sortRes$ix])
    # sort genes by the sort index
    genesSorted <- genes[sortRes$ix]
    # assign genesSorted as the row names of bSorted
    rownames(bSorted) <- genesSorted
    
    # return bSorted
    bSorted <- bSorted
  }

```

Now we can implement the Differntial Genes Analaysis

```{r}
unitV <- chdir(ctrlMat,expmMat,genes)
head(unitV)
```

We got the least of all the DEG's while the sign represents the direction (up or down) and the value represents the magnitued

Now, we want the top 150 upregulated DEG's

```{r}
top_up <- head(unitV[order(unitV[,1], decreasing = T),], n=150)
top_up <- data.frame(top_up)
top_up[,2] <- top_up[,1]
top_up[,1] <- rownames(top_up)
rownames(top_up) <- NULL
write.table(top_up, file="top150.txt", col.names=F, quote=F, row.names=T, sep=", ")
head(top_up)
```

And the top 150 downregulated DEG's

```{r}
top_down <- tail(unitV[order(unitV[,1], decreasing = T),], n=150)
top_down <- data.frame(top_down)
top_down[,2] <- top_down[,1]
top_down[,1] <- rownames(top_down)
rownames(top_down) <- NULL
write.table(top_down, file="btm150.txt", col.names=F, quote=F, row.names=T, sep=", ")
tail(top_down)
```

Now, the list of the genes includeing the values can be insterted into **Enricher** and **L1000CD2**
