# Ichthyosis -  RNA-Seq (STAR + limma/edgeR/DESeq)
This is an analysis of RNA-Seq data of Ichthyosis samples.
Prior, the data was mapped by using STAR.
This code includes quantification (by using featureCounts), basic filteration & normaliztion steps, and differntial expression analysis (by using common toold such as limma, edgeR and DESeq)

```{r include=FALSE}
source ("https://bioconductor.org/biocLite.R")
biocLite()
require(Rsubread)
require(limma)
require(edgeR)
require(DESeq)
require(knitr)
```

## Pre-processing of the data
First, we would like to set the directory and get access to the BAM files which we got after running the STAR mapping

```{r}
knitr::opts_knit$set(root.dir="/Volumes/ShaiM/Netherton/Shai")
bam_files <- list.files(path=".", pattern = ".bam")
```

## Quantification by using featureCounts

```{r}
for (bam_file in bam_files) {
  bam_counted_features <- Rsubread::featureCounts(files = bam_file, 
                                                  annot.ext = "/Users/shai/Desktop/STAR/Homo_sapiens.GRCh38.87.gtf",
                                                  isGTFAnnotationFile = T,
                                                  GTF.featureType = "exon",
                                                  GTF.attrType = "gene_name")
  # Extract the counts object from the list
  extracted_counts <- bam_counted_features$counts
  extracted_counts_genes <- rownames(bam_counted_features$counts)
  extracted_counts_numbers <- bam_counted_features$counts
  dim(bam_counted_features$counts)
  raw_data <- data.frame(extracted_counts_genes, extracted_counts_numbers)
  # Extract sample name
  sample_name <- strsplit(bam_file, '.', fixed=TRUE)[[1]][1]
  # Add .txt extension
  sample_name_txt <- paste0(sample_name, 'extractedCounts.txt')
  # Write the extracted_counts object to the .txt file
  write.table(raw_data, file=sample_name_txt, sep="\t")
}

txt_files <- c(list.files(path=".", pattern="Counts.txt"))
```


## Filtration and normalization steps
We want to prepare the data for the diffrential expression analysis

```{r}
raw_data <- readDGE(txt_files)
```

```{r}
group <- as.factor(c("Ichtyosis","Ichtyosis","Control","Control","Control","Control",
                     "Control","Control","Control","Control","Control","Control","Ichtyosis",
                     "Ichtyosis","Ichtyosis","Ichtyosis","Control","Ichtyosis"))
raw_data$samples$group <- group
```

```{r}
# Check how many genes are not expressed throughout all the samples 
cpm <- cpm(raw_data)
lcpm <- cpm(raw_data, log=T)
```

```{r}
#filter out data
keep.exprs <- rowSums(cpm>1)>=7
raw_data <- raw_data[keep.exprs, keep.lib.sizes=F]
```

```{r}
dim(raw_data)
```


```{r}
colnames(raw_data) <- gsub('_L001_R1_001extractedCounts','',colnames(raw_data))
colnames(raw_data) <- gsub('_L002_R1_001extractedcounts','',colnames(raw_data))
```


In the next step, we will leave R and move to **Python**

This is the code that we will run in Python:
  
  >*import pandas as pd*
  >*def main():*
  >*from clustergrammer import Network*
	>*net = Network()*
	>*net.load_file('big_matrix.txt')*
	>*# net.make_clust(dist_type='cos',views=['N_row_sum', 'N_row_var'])*

	>*net.filter_N_top('row',500, rank_type='sum')*

	>*net.normalize(axis='row', norm_type='zscore')*
	>*net.write_matrix_to_tsv('small_matrix.txt')*
  >*main()*

The `small_matrix.txt` now needs to be uploaded to the **clustergrammer** to check for any clusters

Now we countinue with normalization
```{r}
#normalize data
raw_data <- calcNormFactors(raw_data, method="TMM")
raw_data$samples$norm.factors 

col <- c("red","blue")
col.group <- group
levels(col.group)<- col
col.group <- as.character(col.group)
plotMDS(lcpm,labels = group, col=col.group)
```

## Differential Expression Analysis (Characteristic Direction)

```{r}
design <- model.matrix(~0+group)
colnames(design) <- gsub("group","", colnames(design))
design
# makeContrasts sets the contrasts for pairwise comparisons between the 2 groups
contr.matrix <- makeContrasts(ControlvsIchtyosis = Control- Ichtyosis, levels = colnames(design))
# voom converts raw counts to log-CPM values. experiments with low biological variation
# tend to result in sharp decreasing trend
v <- voom(raw_data, design, plot=T)
# the weight is the relationship between the mean and the variance
# implement linear model so the variance won't be dependent on the mean
# lmfit sets a linear model. It used to fit a separate model to the expression values for each gene
vfit <- lmFit(v, design)
# lmfit & contrast.fit used to fit a separate model to the expression values for each gene
vfit <- contrasts.fit(vfit, contrasts = contr.matrix)
# the empirical Bayas moderation is carried out by borrowing information across all genes to obtain
# more precise estimates of gene-wise variability
efit <- eBayes(vfit)
plotSA(efit)

summary(decideTests(efit))
# treat used to calculate p values from empirical Bayes moderate t-statistics with a minimum log fold
# change requirement
tfit <- treat(vfit, lfc=1)
dt <- decideTests(tfit)
summary(dt)
# now the number of the DEGs was reduced to 107!

v$genes <- rownames(v$E) 
# arrange the genes from smallest to largest adjusted p-value
Control.vs.Ichtyosis <- topTreat(tfit, coef=1, n=Inf)
# highlight the DEGs
plotMD (tfit,column = 1, status=dt[,1], main=colnames(tfit)[1], xlim=c(-8,13))
library(gplots)
contrl.vs.ichtyosis.topgenes <- rownames(Control.vs.Ichtyosis)[1:50]
i <- which(v$genes %in% contrl.vs.ichtyosis.topgenes)
mycol <- colorpanel(1000,"blue","white","red")
par(cex.main=0.5)
```

```{r fig.height=9.0}
heatmap.2 (v$E[i,], scale="row", labRow = rownames(v$E[i,]), labCol = group, trace="none", col=mycol,
           density.info="none", dendrogram = "column", margin=c(5,8), lhei = c(3,8))
```

Now, the list of the genes includeing the values can be insterted into **Enricher** and **L1000CD2**
