---
output:
  html_document: default
  pdf_document: default
---
# RNA-Seq Analysis (Kallisto + limma) - Ichthyosis
RNA-Seq analysis of Ichthyosis samples vs. Control samples.

Prior to the analysis, the data was mapped and quantified by using the Kallisto tool.

The following code includes steps of:
  * Pre-processing steps (gathering the data and filtering steps).
  * Principal Componenet Analysis (PCA).
  * Clusterring the samples by using the Clustergrammer tool.
  * Differntial Expression Analysis by using limma package

```{r include=FALSE}
source ("https://bioconductor.org/biocLite.R")
biocLite()
require("Rsubread")
require("knitr")
require("plyr")
require("reshape")
require("edgeR")
require("DESeq")
require("gplots")
```

## Pre-processing of the data
First, we need to import the human mapping file (GTF file) which will help us to translate the target_ID into gene names. 
```{r}
mapping = "/Users/shai/Downloads/human_mapping.rda"
res = load(mapping)
```

Now, we would like to get access to the files which were obtained by Kallisto. 
```{r}
files = list.files("/Users/shai/Documents/RNASeq/Kallisto/foldercounts", full.names=T)
files
```

Once we have the files, we can run a loop in which we will import the counts from all the files. The counts will be related to the gene level.
```{r}
for(f in files){
  
  # bn will take out all the path from the names of the txt files and will leave only the name of the file
  bn = basename(f)
  
  # abu will read all the files to a table
  abu = read.table(f, sep="\t", stringsAsFactors=F)
  
  # ugene is the col of the genes
  ugene = cb[,2]
  
  # look for number of matches between the gene_id and the gene names
  m3 = match(abu[,1], cb[,1])
  
  # cco is a table with contains only the matches
  cco = cbind(abu,ugene[m3])[-1,]
  
  # co is a table with the counts and gene names only
  co = cco[,c(6,4)]
  co[,1] = as.character(co[,1])
  
  # df is co as a data frame 
  df = data.frame(co[,1], as.numeric(co[,2]))
  colnames(df) = c("gene", "value")
  
  # dd is the data frame with the sum of the counts and numbers or reads
  dd = ddply(df,.(gene),summarize,round(sum=sum(value)),number=length(gene))
  
  ge = dd[,2]
  names(ge) = dd[,1]
  
  write.table(ge, file=paste0("gene_",bn,".tsv"), quote=F, col.names=F, sep="\t")
}
```

Here are the files in which the counts are related to the gene levels:
```{r}
txt_files <- list.files("/Users/shai/Documents/RNASeq/Kallisto/Output/", pattern = "gene_" )
txt_files
```

## Ploting PCA
In order to plot PCA, we will use edgeR & DESeq packages.
```{r}
pca_data <- readDGE(txt_files, path = "/Users/shai/Documents/RNASeq/Kallisto/Output/")
```

Since edgeR & DESeq must have integers in order to implement the PCA steps, we must round the numbers.
```{r}
pca_data$counts <- round(pca_data$counts)
```

```{r}
group <- factor(c("Ichthyosis","Ichthyosis","Control","Control","Control","Control","Control",
                  "Control","Control","Control","Control","Control","Ichthyosis",
                  "Ichthyosis","Ichthyosis","Ichthyosis","Control","Ichthyosis"))
pca_data$samples$group <- group
```

```{r}
cds <- newCountDataSet(pca_data$counts,group )
```

```{r}
cds <- estimateSizeFactors(cds)
```

```{r}
sizeFactors(cds)
```

```{r}
kable(head( counts(cds, normalized=T)))
```

```{r}
cds <- estimateDispersions(cds)
```

```{r}
cdsFull <- newCountDataSet(pca_data$counts, pca_data$samples)
```

```{r}
cdsFull <- estimateSizeFactors(cdsFull)
```

```{r}
cdsFullBlind <- estimateDispersions(cdsFull, method="blind")
```

```{r}
vsdFull <- varianceStabilizingTransformation( cdsFullBlind)
```

```{r}
print(plotPCA(vsdFull, intgroup = "group"))
```


## Filtration and normalization steps
We want to prepare the raw data for the diffrential expression analysis
```{r}
count_matrix <- readDGE(txt_files, path = "/Users/shai/Documents/RNASeq/Kallisto/Output/")
count_matrix$counts <- round(count_matrix$counts)
group <- as.factor(c("Ichtyosis","Ichtyosis","Control","Control","Control","Control",
                     "Control","Control","Control","Control","Control","Control","Ichtyosis",
                     "Ichtyosis","Ichtyosis","Ichtyosis","Control","Ichtyosis"))
count_matrix$samples$group <- group
colnames(count_matrix) <- gsub('_L001_R1_001','',colnames(count_matrix))
colnames(count_matrix) <- gsub('_L002_R1_001','',colnames(count_matrix))
samplenames <- rownames(count_matrix$samples)
cpm <- cpm(count_matrix)
lcpm <- cpm(count_matrix, log=T)
```

```{r}
# check if there are genes with no expression at all
table(rowSums(count_matrix$counts==0)==18)
```

```{r}
#filter out data
keep.exprs <- rowSums(cpm>1)>=6
count_matrix <- count_matrix[keep.exprs, keep.lib.sizes=F]
dim(count_matrix)
```

```{r}
count_matrix <- calcNormFactors(count_matrix, method="TMM")
count_matrix$samples$norm.factors
```

## Clustergrammer
```{r}
cluster_grammer_data <- count_matrix$counts
colnames(cluster_grammer_data) <- group
write.table(cluster_grammer_data, file='big_matrix.txt', sep="\t", quote = F)
```

In this step, we will leave R and move to **Python**. We will upload the `big_matrix.txt` file and will derive the `small_matrix.txt` file. The `small_matrix.txt` file will be uploaded to the clustergrammer.

This is the code that we will run in Python:
  `import pandas as pd`
  `def main():`
  `from clustergrammer import Network`
	`net = Network()`
	`net.load_file('big_matrix.txt')`
	`# net.make_clust(dist_type='cos',views=['N_row_sum', 'N_row_var'])`
	`net.filter_N_top('row',500, rank_type='sum')`
	`net.normalize(axis='row', norm_type='zscore')`
	`net.write_matrix_to_tsv('small_matrix.txt')`
  `main()`
  
  
  
## Differential expression analysis

Creating a design matrix
```{r}
design <- model.matrix(~0+group)
colnames(design) <- gsub("group","",colnames(design))
kable(design)
```

Creating a contrasts matrix
```{r}
contr.matrix <- makeContrasts(Control_vs_Ichtyosis= Control-Ichtyosis, levels = colnames(design))
kable(contr.matrix)
```

Plot the mean-variance of log cpm values
```{r}
v <- voom(count_matrix, design, plot=T)
```

```{r}
v
```

Fitting linear model
```{r}
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts = contr.matrix)
efit <- eBayes(vfit)
plotSA(efit)
```

Examining the numbers of the DE genes
```{r}
kable(summary(decideTests(efit)))
```

```{r}
tfit <- treat(vfit, lfc=1)
dt <- decideTests(tfit)
kable(summary(dt))
```

```{r}
Control.vs.Ichtyosis <- topTreat(tfit, coef=1, n=Inf)
kable(head(Control.vs.Ichtyosis))
```

```{r}
plotMD(tfit, column=1, status=dt[,1], main=colnames(tfit)[1], xlim=c(-8, 13))
```

```{r}
Control.vs.Ichtyosis.topgenes <- rownames(Control.vs.Ichtyosis)[1:50]
i <- which(rownames(v$E) %in% Control.vs.Ichtyosis.topgenes)
mycol <- colorpanel(1000,"blue","white","red")
par(mar=c(5,4,4,2))
heatmap.2(v$E[i,],
          #scale="row",
          labRow = rownames(v$E)[i],
          labCol = group,
          col=mycol, 
          trace="none",
          density.info = "none",
          margin=c(5,8),
          lhei=c(5,13),
          dendrogram = "column",
          #key.xlab = 'log-cpm',
          key.title='',
          keysize = 1,
          cexRow = 0.5)
```